### ISSUE TITLE
ask for jack settings / save them once in the session and restore them when the session is loading

### ISSUE THREAD

**Issue by newlaurent62** [Houston4444/RaySession#64]
Hello !

In my wizards, I would like to add an optional wizard page asking for  Jack user settings for the created session:
- Interface, 
- Frames/Period 
- and Sample rate.

After validating the wizard inputs, it will store them in the Raysession path. The desired Jack settings will be loaded with the Raysession loading. (using jack_control)

Do you have something similar already implemented in raysession or that i can use to have this feature ready ? Any thought ?

BR,
Laurent

**Comment by Houston4444** [comment]
Yes.
That's exactly what jack_config session scripts does.
It saves and reloads all set jack parameters.
jack parameters are saved in plain/text jack_parameters file in session folder.
So if you want a session to load specific jack parameters, just write the parameters you wants in this file.
hostname is also saved in this file, this way, if session is copied and loaded on another machine, it will only cares about samplerate (and not interface).

**Comment by newlaurent62** [comment]
I started jack on my internal sound card (the only sound card on the machine)

I created a new session with jack_config session template.

I saved the session and look at the .jack_config file in the ray-scripts dir but nothing has been append to the file, no parameters.

When i close the session, it complains: "L'interface hw:PCH n'est pas connectée, votre ancienne configuration n'a pas pu être restaurée"

Je ne comprends pas le message d'erreur.

`cat /proc/asound/cards`:
` 0 [PCH            ]: HDA-Intel - HDA Intel PCH
                      HDA Intel PCH at 0xdf220000 irq 131
`

`aplay --list-devices`
```
`**** Liste des Périphériques Matériels PLAYBACK ****
carte 0: PCH [HDA Intel PCH], périphérique 0: ALC269VC Analog [ALC269VC Analog]
  Sous-périphériques: 0/1
  Sous-périphérique #0: subdevice #0
carte 0: PCH [HDA Intel PCH], périphérique 3: HDMI 0 [HDMI 0]
  Sous-périphériques: 1/1
  Sous-périphérique #0: subdevice #0
carte 0: PCH [HDA Intel PCH], périphérique 7: HDMI 1 [HDMI 1]
  Sous-périphériques: 1/1
  Sous-périphérique #0: subdevice #0
carte 0: PCH [HDA Intel PCH], périphérique 8: HDMI 2 [HDMI 2]
  Sous-périphériques: 1/1
  Sous-périphérique #0: subdevice #0
carte 0: PCH [HDA Intel PCH], périphérique 9: HDMI 3 [HDMI 3]
  Sous-périphériques: 1/1
  Sous-périphérique #0: subdevice #0
carte 0: PCH [HDA Intel PCH], périphérique 10: HDMI 4 [HDMI 4]
  Sous-périphériques: 1/1
  Sous-périphérique #0: subdevice #0
```

et `arecord --list-devices` renvoie aussi un résultat.

```
carte 0: PCH [HDA Intel PCH], périphérique 0: ALC269VC Analog [ALC269VC Analog]
  Sous-périphériques: 0/1
  Sous-périphérique #0: subdevice #0
```

**Comment by Houston4444** [comment]
Ça c'est bizarre.
peux-tu me renvoyer la sortie de la commande suivante ?
`LANG=C LC_ALL=C aplay -l`

**Comment by Houston4444** [comment]
> I saved the session and look at the .jack_config file in the ray-scripts dir but nothing has been append to the file, no parameters.

This is normal. JACK parameters are saved in file **jack_parameters** in the session dir.

**Comment by newlaurent62** [comment]
ok thanks i found the jack_parameters file in the session dir when i click on save button.

Here is the output of `LANG=C LC_ALL=C aplay -l`
```
**** List of PLAYBACK Hardware Devices ****
card 0: PCH [HDA Intel PCH], device 0: ALC269VC Analog [ALC269VC Analog]
  Subdevices: 0/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 3: HDMI 0 [HDMI 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 7: HDMI 1 [HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 8: HDMI 2 [HDMI 2]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 9: HDMI 3 [HDMI 3]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: PCH [HDA Intel PCH], device 10: HDMI 4 [HDMI 4]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

**Comment by newlaurent62** [comment]
> hostname is also saved in this file, this way, if session is copied and loaded on another machine, it will only cares about samplerate (and not interface).

What about a USB sound card  ? Will you have to reconfigure it on each machine or will it detect that the sound card is the same and go on ?

**Comment by newlaurent62** [comment]
En plus du message ci-dessus:

> This is normal. JACK parameters are saved in file jack_parameters in the session dir.

Why the jack_parameters filename is not preceed by the session name ?

**Comment by Houston4444** [comment]
Je vais investiguer ce bug.

> What about a USB sound card ? Will you have to reconfigure it on each machine or will it detect that the sound card is the same and go on ?

No, it won't. it's a little arbitrary, but in almost cases, this interface won't exists (case session is drop from one user to another one).
There are some choices a little arbitrary in this script, because it's very hard to know what user really wants, maybe the script could have options with env variables. The main part of the script is in src/jack_config_script/ . Feel free to propose code modifications !

> Why the jack_parameters filename is not preceed by the session name ?

Simply because this file has no link with any client of the session !

**Comment by newlaurent62** [comment]
I tested with two sound cards on the machine. I selected a usb sound card rather than the internal sound card and the message is the same but for the selected sound card.

**Comment by Houston4444** [comment]
Ok, I saw the bug too.
JACK is now restarting the first time (that was wanted), and now I do not see this bug anymore at session close.

JACK is restarting the first time because JACK doesn't provide any reliable way to get the real parameters of the running driver, so a python dbus daemon (ray-jack_checker_daemon) is launched to check JACK parameters at JACK start. If JACK is already started, it informs that jack parameters are not reliable. Once this little daemon is started, JACK is not restarted at session load.

The idea is to purpose to start this daemon at desktop startup when creating a session from jack_config template (in a dialog window with an non-again checkbox). Of course, the best would be that JACK itself write a little file with its parameters when it starts or switch_master. Waiting for this, this script can be useful.

**Comment by newlaurent62** [comment]
i check the daemon and there is a broken link at least on my machine:

`ray-jack_checker_daemon -> /home/houston/RaySession/src/jack_config_script/jack_checker_daemon.py
`

**Comment by newlaurent62** [comment]
I started the daemon. It seemed to run but print nothing.

below you will find two screenshots. One with qjacktl started and another with qjackctl stopped before starting the session.

![raysession-screenshot2](https://user-images.githubusercontent.com/10225394/83463557-aa600a00-a46e-11ea-860b-9b06f65e9512.png)
![raysession-screenshot1](https://user-images.githubusercontent.com/10225394/83463559-aaf8a080-a46e-11ea-9da3-f484392ed481.png)

**Comment by Houston4444** [comment]
> i check the daemon and there is a broken link at least on my machine:

Ah, sorry, it is now fixed.

Mmmmmmhh, have you update your repo ?

**Comment by newlaurent62** [comment]
Yes i did:
This morning i get two files:
```
Mise à jour 433be01..496fdab
Fast-forward
 src/bin/ray-jack_checker_daemon | 2 +-
 src/daemon/session.py           | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

```

```
git pull
sudo make uninstall
sudo make install
```

**Comment by Houston4444** [comment]
Ok, I think I find it now.
Probably a problem with jack interface name which can contains ',' or not.
Please pull and retry, sorry.
Désolé de te faire essuyer les plâtres.

**Comment by newlaurent62** [comment]
Y a pas de souci. Je suis développeur je sais ce que c'est.

Je teste ça ce soir à mon retour.

**Comment by newlaurent62** [comment]
Ok it works.

By the way, if i configure a session to use a USB card and don't plug it in the system before loading session, il doesn't ask me to do so.

**Comment by newlaurent62** [comment]
May i also suggest to allow to set a sound card without hostname or generic hostname like localhost or 127.0.0.1 to let the user tell RaySession that this jack setup has to be used on every machine ?

**Comment by Houston4444** [comment]
> By the way, if i configure a session to use a USB card and don't plug it in the system before loading session, il doesn't ask me to do so.

The normal thing is just to say that this is not possible. By design, the script never ask the user anything, just give some informations about quite long process, and session load fail.
Because one of the interest of ray_control is to can do all session operations without the GUI. There are a lot of use-cases where the GUI is just a too much window (work is mostly done with DAWs and hosts). without GUI, a user ask could become an always waiting action.

> May i also suggest to allow to set a sound card without hostname or with a generic hostname like localhost or 127.0.0.1 to let the user tell to RaySession that this jack setup has to be used on every machine ?

for this, I didn't test, but I think that few lines to remove the line ^hostname: in jack_parameters in save.sh will do the job.

**Comment by newlaurent62** [comment]
I understand your point of view however a session load without sound card connected when it is needed leads to cpu and time waste.

The default behavior should be to pass and ignore errors while the user does not require software check explicitly.

In my use case, i will face noobs users and i prefer to prevent errors rather than letting them react to software failures ...

Would you like to consider creating a more restrictive check in addition to the ignore errors mode by asking the end user what is needed ? (The gui could be replace with terminal input for those using ray_control only)

**Comment by Houston4444** [comment]
> The gui could be replace with terminal input for those using ray_control only

No, because ray_control could be launched in a script (for example at desktop session startup). There is a ray_control option to display a question dialog (`ray_control script_user_action TEXT`), maybe the good behavior should be to return an error code when no gui is connected to the daemon. The thing with this is that we never end the script possibilities (why not combobox, edit line or whatever...), and I don't want to rewrite kdialog or zenity, they exists ;)

Also, as possible I want this script to be able to work on a non graphical environment.

If you want you can easily fork the script and change its behavior using dialog windows (with kdialog, zenity, or self made windows), and I am absolutely not opposite to help you.
Note: Maybe you guessed it, the main part of the script is in RS code and not directly in the session scripts dir, this is to can update this part with RS update.

**Comment by Houston4444** [comment]
> The default behavior should be to pass and ignore errors while the user does not require software check explicitly.

The problem is that absence of required sound card will result in a jack start failed, so I disagree.
But disable theses checks with an environment variable (or option) exported in the session scripts dir could be quickly implemented.

**Comment by newlaurent62** [comment]
> There is a ray_control option to display a question dialog (ray_control script_user_action TEXT), maybe the good behavior should be to return an error code when no gui is connected to the daemon. 

I already use it in my load script to ask a question (I inform the user about the error,tell him what action to do and ask him if he wants to stop loading or ignore the error):

- if a program is missing i ask a question
- if i detect that a needed server is not started, i ask question
- if i detect that jack is not running, i ask question
- if additionnal audio devices (alsa_in/alsa_out) are missing, i ask a question

It is unlikely that the user answers every questions at session load ... 
My load script uses .env file in ray-scripts dir to set the ray-scripts behavior. You can disable every check made by the ray-scripts

**Comment by Houston4444** [comment]
Do you need a global no-check or only for alsa ?
Currently checks only concerns alsa and net drivers.
firewire seems not so easy to check.

**Comment by newlaurent62** [comment]
My Jack use case is simple:

1. Stop jack server if any
2. start jack server whatever the selected driver is (load parameters from jack_parameters file)
alsa_in and alsa_out can be used with net drivers i think (while i'm not sure) 

Jack_parameters are set with the wizard (so they will never be changed during the session). If the user want to change parameters. he has three solutions:

- change it in the jack parameters file directly
- create another session 
- run wizard to change jack parameters (not yet implemented)

**Comment by newlaurent62** [comment]
But my main target is alsa driver.

I will do various check on top of your script.

**load**
Check if the jack_parameters device is available on the system. If not ask question (Stop or ignore)
if error and ignore and jack is started don't stop jack (we need to use a jack server for the session) 
if error and ignore and jack is stopped, we run the steps ...
if no error then stop jack, set new jack parameters, restart jack

**save**
If device is missing, is it likely to have software failures here ? If yes, then i will ask the user to start jack server before any save.

**close**
if the device is missing and jackdbus runs timeout then kill jackdbus and restart it

Any thought ?

**Comment by Houston4444** [comment]
Hi !

> Stop jack server if any

Quite shame, the goal of the ray-jack_checker_daemon is to avoid this when not needed.

> alsa_in and alsa_out can be used with net drivers i think (while i'm not sure)

Sure it is possible, it is the main use case of theses 2 commands (Note there is a little more latency on the alsa_out jack ports).

> Jack_parameters are set with the wizard (so they will never be changed during the session).

Perfectly understandable. Just remove save.sh if you don't need it for something else.

**Comment by Houston4444** [comment]
> **load**
> Check if the jack_parameters device is available on the system. If not, ask question (Stop loading or ignore)

Ok, I'll make an env var to disable check devices.
Do you want an executable to use the same check_device method as the original script ? The advantage is to can get all updates if any.

> if error and ignore and jack is started don't stop jack (we need to use a jack server for the session)
> if error and ignore and jack is stopped, we run the steps ...

If the loading session has already jack_started:1 in jack_parameters, that is the default behavior of the script.

> if no error then stop jack, set new jack parameters, restart jack

Quite shame as I said in the previous message. Be careful with jack stop at session load, don't forget to make a `ray_control clear_clients` in case session is switching from another session.

> save
> If device is missing, will it likely to have software failures here ?

What do you want in this case ? Save and abort with message ?

> **close**
> if the device is missing and jackdbus runs timeout then kill jackdbus and restart it

Which device are you talking about ? Device used for the session or device to restore ?
How do you check jackdbus runs timeout ?

**Comment by newlaurent62** [comment]
The reason why i would restart (stop and start) the jack server is, if any jack client created before the session load has a jack client name already used by the RaySession clients, the session will not load as expected especially jack patch ...

So to prevent patch break in the jack patch, i have to be sure that no jack client has been started with jack client name already used by the RaySession clients loaded.

There is perhaps something more clever to do ...

**Comment by newlaurent62** [comment]
> Ok, I'll make an env var to disable check devices.
> Do you want an executable to use the same check_device method as the original script ? The 
> advantage is to can get all updates if any.

Good ! If you can adapt your script rather than doing a fork, it is better. I've already implemented in my wizard the writing of the jack_parameters file. 

> Quite shame as I said in the previous message. Be careful with jack stop at session load, don't 
forget to make a ray_control clear_clients in case session is switching from another session.

See my previous post for an explanation

> What do you want in this case ? Save and abort with message ?

Well here, there's a logic error. We check that jack server is running rather than missing device. 
I guess that if jack server is not running, most of the jack audio applications will have save failure ... so we abort save when jack is not running.

> Which device are you talking about ? Device used for the session or device to restore ?

Ideally we should check the device set at the session opening. 

> How do you check jackdbus runs timeout ?

Simply by running  jack_control stop with a timeout ? (i already use call with timeout in one of my bash script)

**Comment by Houston4444** [comment]
> The reason why i would restart (stop and start) the jack server is, if any jack client created before the session load has a jack client name already used by the RaySession clients, the session will not load as expected especially jack patch ...

It's kind of relevant, but I won't happen often in fact. This could happen only when we have programs already launched out of RS before session load, and when these programs are also present in the session. I have to say I started from the premise that user launch all its JACK programs inside the session (except utilities as PulseAudio, a2j...), and I think it is the good way user has to do. I suggest a non-again dialog at start of your wizard session make saying this, and also I'll add it to the non-again dialog at session creation from jack_config template (not already implemented).

>  ... so we abort save when jack is not running.

humm, not sure, maybe some clients have not been saved for a long. I think almost clients will save kindly anyway if we ask to, even in this situation.

Your proposition to check device and jackdbus responding is good, I will implement it.
Thanks you for the `timeout` command, I didn't know this command ;)

**Comment by newlaurent62** [comment]
Hello,

Are you ok with this ?

**load**:
- if jack is started and check env set to on : check the device is the one in jack_parameters otherwise return an error so that the scipt can ask a question to ignore or continue
- if jack not started and check env set to on : return an error so that the main load.sh can ask a question to ignore or continue (and give the user the opportunity to start jack)

**save** : 
- I will save anyway even if jack is not started ...

**close**:
- if the device is not the same as the one we had at the session start, raise an error so that the close.sh can ask question or react to this state. 
- If the device is missing and jack_control stop runs timeout kill jackdbus ...

Another feature i would need:

In some cases, the jack settings must fill some prerequisites:
for example: 48000 Hz between 64 & 512 frames ...

Could you add constraints check to your script ?

BR,
Laurent

**Comment by Houston4444** [comment]
Hello !

> if jack is started and check env set to on : check the device is the one in jack_parameters otherwise return an error so that the script can ask a question to ignore or continue

Strange. So, only the device ? Continue if device is the same and samplerate different ? by continue, I mean "let the script stop/restart JACK with the good parameters".

> if jack not started and check env set to on : return an error so that the main load.sh can ask a question to ignore or continue (and give the user the opportunity to start jack)

Probably possible.

> if the device is not the same as the one we had at the session start, raise an error so that the close.sh can ask question or react to this state.

mmmh, your close.sh could directly read the jack parameters. the daemon ray-jack_checker_daemon writes the file `/tmp/RaySession/jack_current_parameters`

> If the device is missing and jack_control stop runs timeout kill jackdbus ...

Ok by default.

> In some cases, the jack settings must fill some prerequisites:
> for example: 48000 Hz between 64 & 512 frames ...
> 
> Could you add constraints check to your script ?

No, too complex to know which constraints an user could want. But you can do it, still reading directly `/tmp/RaySession/jack_current_parameters` (if the file doesn't exists, start ray-jack_checker_daemon). Note that the JACK parameters before first session start are saved in `/tmp/RaySession/jack_backup_parameters`.

I will probably add one env variable for all the available options, such as RAY_JACK_SCRIPT_OPTIONS=":manage_pulseaudio:force_jack_start:etc:"

**Comment by newlaurent62** [comment]
> if jack is started and check env set to on : check the device is the one in jack_parameters otherwise return an error so that the script can ask a question to ignore or continue

You are right when i say check device i mean check jack_parameters are the same (at least device, frames, sample rate).

> No, too complex to know which constraints an user could want. 

Indeed, what i ask for is a constraint parameter file that would be filled by the wizard if needed.

> I will probably add one env variable for all the available options, such as 
RAY_JACK_SCRIPT_OPTIONS=":manage_pulseaudio:force_jack_start:etc:"

Please let me know when you're done.

**Comment by Houston4444** [comment]
I 've set the following env vars, they all have to be 'true' or 'false'

RAY_MANAGE_PULSEAUDIO
default:true
reconfigure pulseaudio bridge with number of jack Source/Sink ports after jack restart

RAY_JACK_RELIABILITY_CHECK
default:true
if false, consider the current parameters given by jackdbus as reliable,that means JACK can be not restarted if the ray-jack_checker_daemon is not started

RAY_HOSTNAME_SENSIBLE
default:true
if false, do not save hostname in jack_parameters and behaves as if there was no hostname difference at load.

RAY_FAIL_IF_JACK_DIFF
default:false
exit with error code 29 (why not 29 ? ;) ), when there is a wanted parameter not the same as the current one (also exit 29 at save if jack current parameters are different to theses ones on the jack_parameters file).

I'll also add something to manage if the script has to start/stop JACK if jack is marked as stopped in jack_parameters session file, and in the case jack_parameters is not existant.

I think to something, instead to run the script with RAY_FAIL_IF_JACK_DIFF, I could add an option to `ray-jack_config_script` to get diff_parameters, then you could display to user what is wrong with the current JACK parameters. I could also easily add an option to set JACK parameters.

**Comment by newlaurent62** [comment]
> I think to something, instead to run the script with RAY_FAIL_IF_JACK_DIFF, I could add an option to ray-jack_config_script to get diff_parameters, then you could display to user what is wrong with the current JACK parameters. I could also easily add an option to set JACK parameters.

Yes that would be nice to have. (diff and set).

**Comment by Houston4444** [comment]
ok that is done.
`ray-jack_config_script get_diff`
returns a multi-lines string with all parameters that need to be changed to open session.
Optionnaly with a multi-lines arguments, it will gives diff between string in arguments and current parameters.

`ray-jack_config_script set_jack_parameters ARGUMENT`
ARGUMENT has to be a multi-lines argument with the same look as jack_parameters session file.
It will set jack parameters in argument.

Can I remove RAY_FAIL_IF_JACK_DIFF ?

**Comment by newlaurent62** [comment]
I have not started to implement the use of new ray-jack_config_script features ... so Yes, you can remove it.

**Comment by newlaurent62** [comment]
Thanks a lot ! 

I tested different load cases and it seems to work fine.

on save, at least, period and frames are not saved when different from the jack settings of the ray session (but device is saved if different). Is it normal ?

`
RAY_HOSTNAME_SENSIBLE=false ` works fine,
`
ray-jack_config_script  get_diff ` works fine too at least for device, frames and period.

**Comment by Houston4444** [comment]
> on save, at least, period and frames are not saved when different from the jack settings of the ray session (but device is saved if different). Is it normal ?

No it is not, I really don't see what could make a behavior difference between device and other parameters.

**Comment by newlaurent62** [comment]
Ok sorry, i made a mistake. That works well.

Thanks a lot !
